doc/design/dist/README.md

```diff
@@ -0,0 +1,190 @@
+
+- Master process
+- Trainer process
+- Parameter server process
+
+One training job will only have one master process, typically multiple trainer processes and parameter server processes. Their relation is illustrated in the following graph:
+
+<img src="src/paddle-model-sharding.png" height="400"/>
+
+### Master Process
+
+The master process will:
+
+- Do [parameter server selection](#parameter-server-selection).
+- Shard dataset into [tasks](#task) and dispatch tasks to trainers.
+- Keep track of training progress on the dataset with [task queue](#task-queue). A training job will iterate on the dataset for a full pass until it goes into next pass.
+
+Now we will explain the concepts mentioned above:
+
+#### Selection Request
+
+The selection request is a request that the master sends to a parameter server candidate, making it a parameter server available to the trainers. It contains information such as the parameter server index, the optimizaiton algorithm, the parameter save period, and the path for saving parameters.
+
+#### Parameter Server Selection
+
+The parameter server selection process selects parameter servers from parameter server candidates. It ensures the parameter servers that the trainers see are in consistent order, since the trainer needs to decide the parameter placement according to the consistent order.
+
+The selection process is as follows:
+
+- The master watches `/ps_candidate/` prefix in etcd. When a parameter server candidate joins and there is not enough parameter servers, the master will remove the candidate's entry in `/ps_candidate/` and send a [selection reqeust](#selection-request) to the candidate. Upon receiving the request, the candidate will set key `/ps/<index>` in etcd with a lease to make itself available for the trainers. The `<index>` is from the selection request.
+
+	The graph below shows a parameter server candidate come online and then being selected, available for the trainers:
+	
+	<img src="src/paddle-ps-0-can.png" width="650"/>
+	<img src="src/paddle-ps-0-sel.png" width="650"/>
+
+- The master watches `/ps/` prefix in etcd. When a selected parameter server went offline, the master will select a not yet selected parameter server candidate by sending the selection request to fill the missing parameter server spot.
+
+	The graph below shows one parameter server is missing, the cluster management system created a new parameter server. The new parameter server announced itself as a candidate. Then the master filled the missing parameter server spot with the new candidate.
+
+	<img src="src/paddle-ps-new-can.png" width="650"/>
+	<img src="src/paddle-ps-new-sel.png" width="650"/>
+
+
+#### Task 
+
+A task is a piece of sharded data to be trained. The total number of tasks will be much bigger than the total number of trainers. The number of data instances inside a task will be much bigger than the mini-batch size.
+
+#### Task Queue
```

Comment:
@helinwang starting jobs with more than 10000 shards of data, when pushing shards(task) to todo queue, may cause too many writes to etcd. keep the todo queue in master's memory may solve this, etcd will handle only a small amount of writes. This will increase the maximum concurrent jobs on the cluster since we only deploy one etcd cluster.